File: app/controllers/api/v1/base_controller.rb
module Api
  module V1
    class BaseController < ActionController::API
      include ActionController::HttpAuthentication::Token::ControllerMethods

      before_action :authenticate_api_token!

      rescue_from ActiveRecord::RecordNotFound, with: :not_found
      rescue_from ActiveRecord::RecordInvalid, with: :unprocessable_entity
      rescue_from ActionController::ParameterMissing, with: :bad_request

      private

      def authenticate_api_token!
        authenticate_or_request_with_http_token do |token, _options|
          @current_user = AdminUser.find_by(api_token: token)
        end
      end

      attr_reader :current_user

      def not_found(error)
        render json: { error: 'Not Found', message: error.message }, status: :not_found
      end

      def unprocessable_entity(error)
        render json: { error: 'Unprocessable Entity', errors: error.record.errors.full_messages },
               status: :unprocessable_entity
      end

      def bad_request(error)
        render json: { error: 'Bad Request', message: error.message }, status: :bad_request
      end
    end
  end
end

File: laravel-app/app/Http/Controllers/Api/V1/BaseController.php
<?php

namespace App\Http\Controllers\Api\V1;

use App\Http\Controllers\Controller;
use App\Models\AdminUser;
use Illuminate\Http\Request;
use Illuminate\Database\Eloquent\ModelNotFoundException;
use Illuminate\Validation\ValidationException;
use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;

class BaseController extends Controller
{
    /**
     * The authenticated API user
     */
    protected ?AdminUser $currentUser = null;

    /**
     * Create a new controller instance
     */
    public function __construct()
    {
        $this->middleware(function (Request $request, $next) {
            $this->authenticateApiToken($request);
            return $next($request);
        });
    }

    /**
     * Authenticate API token from Authorization header
     */
    protected function authenticateApiToken(Request $request): void
    {
        $token = $request->bearerToken();

        if (!$token) {
            abort(401, 'Unauthorized: Missing API token');
        }

        $this->currentUser = AdminUser::where('api_token', $token)->first();

        if (!$this->currentUser) {
            abort(401, 'Unauthorized: Invalid API token');
        }
    }

    /**
     * Get the authenticated user
     */
    protected function getCurrentUser(): ?AdminUser
    {
        return $this->currentUser;
    }

    /**
     * Handle model not found exceptions
     */
    protected function handleNotFound(\Exception $e)
    {
        return response()->json([
            'error' => 'Not Found',
            'message' => $e->getMessage(),
        ], 404);
    }

    /**
     * Handle validation exceptions
     */
    protected function handleValidationException(ValidationException $e)
    {
        return response()->json([
            'error' => 'Unprocessable Entity',
            'errors' => $e->errors(),
        ], 422);
    }

    /**
     * Handle bad request exceptions
     */
    protected function handleBadRequest(\Exception $e)
    {
        return response()->json([
            'error' => 'Bad Request',
            'message' => $e->getMessage(),
        ], 400);
    }
}
